File: src/oca/java/src/org/opennebula/client/PoolElement.java
/*******************************************************************************
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client;

import java.io.ByteArrayInputStream;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Document;
import org.w3c.dom.Node;

/**
 * Represents a generic element of a Pool in
 * XML format.
 *
 */
public abstract class PoolElement {

    protected static XPath xpath;

    protected int    id;
    protected Node   xml;

    protected Client client;

    /**
     * Creates a new PoolElement with the specified attributes.
     * @param id Id of the element.
     * @param client XML-RPC Client.
     */
    protected PoolElement(int id, Client client)
    {
        if(xpath == null)
        {
            XPathFactory factory = XPathFactory.newInstance();
            xpath = factory.newXPath();
        }

        this.id     = id;
        this.client = client;
    }

    /**
     * Creates a new PoolElement from the xml provided.
     *
     * @param client XML-RPC Client.
     * @param xmlElement XML representation of the element.
     */
    protected PoolElement(Node xmlElement, Client client)
    {
        if(xpath == null)
        {
            XPathFactory factory = XPathFactory.newInstance();
            xpath = factory.newXPath();
        }

        this.xml    = xmlElement;
        this.client = client;
        this.id     = Integer.parseInt(xpath("ID"));
    }

    /**
     * After a *.info call, this method builds the internal xml
     * representation of the pool.
     * @param info The XML-RPC *.info response
     */
    protected void processInfo(OneResponse info)
    {
        if (info.isError())
        {
            return;
        }

        try
        {
            DocumentBuilder builder =
                DocumentBuilderFactory.newInstance().newDocumentBuilder();
            Document doc = builder.parse(
                new ByteArrayInputStream(info.getMessage().getBytes("UTF-8")));

            xml = doc.getDocumentElement();
        }
        catch (Exception e) {}
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param method XML-RPC method.
     * @param id The id of the target object.
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    protected static OneResponse chmod(Client client, String method, int id,
                                    int owner_u, int owner_m, int owner_a,
                                    int group_u, int group_m, int group_a,
                                    int other_u, int other_m, int other_a)
    {
        return client.call(method, id,
                            owner_u, owner_m, owner_a,
                            group_u, group_m, group_a,
                            other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param method XML-RPC method.
     * @param id The id of the target object.
     * @param octet Permissions octet, e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    protected static OneResponse chmod(Client client, String method, int id,
                                        String octet)
    {
        int owner_u = (Integer.parseInt(octet.substring(0, 1)) & 4) != 0 ? 1 : 0;
        int owner_m = (Integer.parseInt(octet.substring(0, 1)) & 2) != 0 ? 1 : 0;
        int owner_a = (Integer.parseInt(octet.substring(0, 1)) & 1) != 0 ? 1 : 0;
        int group_u = (Integer.parseInt(octet.substring(1, 2)) & 4) != 0 ? 1 : 0;
        int group_m = (Integer.parseInt(octet.substring(1, 2)) & 2) != 0 ? 1 : 0;
        int group_a = (Integer.parseInt(octet.substring(1, 2)) & 1) != 0 ? 1 : 0;
        int other_u = (Integer.parseInt(octet.substring(2, 3)) & 4) != 0 ? 1 : 0;
        int other_m = (Integer.parseInt(octet.substring(2, 3)) & 2) != 0 ? 1 : 0;
        int other_a = (Integer.parseInt(octet.substring(2, 3)) & 1) != 0 ? 1 : 0;

        return chmod(client, method, id,
                owner_u, owner_m, owner_a,
                group_u, group_m, group_a,
                other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param client XML-RPC Client.
     * @param method XML-RPC method.
     * @param id The id of the target object.
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    protected static OneResponse chmod(Client client, String method, int id,
                                        int octet)
    {
        return chmod(client, method, id, Integer.toString(octet));
    }

    /**
     * Returns the element's ID.
     * @return the element's ID.
     */
    public String getId()
    {
      return Integer.toString(id);
    }

    public int id()
    {
        return id;
    }

    /**
     * Returns the element's name.
     * @return the element's name.
     */
    public String getName()
    {
        return xpath("NAME");
    }

    /**
     * Performs an xpath evaluation for the "state" expression.
     * @return The value of the STATE element.
     */
    public int state()
    {
        String state = xpath("STATE");

        return state != null ? Integer.parseInt( state ) : -1;
    }

    /**
     * Returns the owner User's ID, or -1 if the element doesn't have one.
     *
     * @return the owner User's ID, or -1 if the element doesn't have one.
     */
    public int uid()
    {
        String  uid_str = xpath("UID");
        int     uid_int = -1;

        if ( uid_str != null )
        {
            try
            {
                uid_int = Integer.parseInt( uid_str );
            }
            catch (NumberFormatException e) {}
        }

        return uid_int;
    }

    /**
     * Returns the element group's ID, or -1 if the element doesn't have one.
     *
     * @return the element group's ID, or -1 if the element doesn't have one.
     */
    public int gid()
    {
        String  gid_str = xpath("GID");
        int     gid_int = -1;

        if ( gid_str != null )
        {
            try
            {
                gid_int = Integer.parseInt( gid_str );
            }
            catch (NumberFormatException e) {}
        }

        return gid_int;
    }

    /**
     * Evaluates an XPath expression and returns the result as a String.
     * If the internal xml representation is not built, returns null. The
     * subclass method info() must be called before.
     *
     * @param expression The XPath expression.
     * @return The String that is the result of evaluating the
     * expression and converting the result to a String. An empty String is
     * returned if the expression is not a valid path; null if
     * the internal xml representation is not built.
     */
    public String xpath(String expression)
    {
        String result = null;

        try
        {
          result = xpath.evaluate(expression, xml);
        }
        catch (XPathExpressionException e) {}

        return result;
    }
}

File: src/oca/ruby/opennebula/pool_element.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2023, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'opennebula/pool'

module OpenNebula

    # The PoolElement Class represents a generic element of a Pool in
    # XML format
    class PoolElement < XMLElement

        protected

        # node:: _XML_is a XML element that represents the Pool element
        # client:: _Client_ represents a XML-RPC connection
        def initialize(node, client)
            @xml    = node
            @client = client

            if self['ID']
                @pe_id = self['ID'].to_i
            else
                @pe_id = nil
            end
            @name = self['NAME'] if self['NAME']
        end

        #######################################################################
        # Common XML-RPC Methods for all the Pool Element Types
        #######################################################################

        # Common client call wrapper. Checks that @pe_id is defined, and
        # returns nil instead of the response if it is successful
        #
        # @param [String] xml_method xml-rpc method
        # @param [Array] args any arguments for the xml-rpc method
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def call(xml_method, *args)
            return Error.new('ID not defined') unless @pe_id

            rc = @client.call(xml_method, *args)
            rc = nil unless OpenNebula.is_error?(rc)

            rc
        end

        # Calls to the corresponding info method to retreive the element
        # detailed information in XML format
        #
        # @param [String] xml_method the name of the XML-RPC method
        # @param [String] root_element Base XML element name
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def info(xml_method, root_element, decrypt = false)
            return Error.new('ID not defined') unless @pe_id

            rc = @client.call(xml_method, @pe_id, decrypt)

            if !OpenNebula.is_error?(rc)
                initialize_xml(rc, root_element)
                rc = nil

                @pe_id = self['ID'].to_i if self['ID']
                @name  = self['NAME'] if self['NAME']
            end

            rc
        end

        # Calls to the corresponding allocate method to create a new element
        # in the OpenNebula core
        #
        # @param [String] xml_method the name of the XML-RPC method
        # @param [Array] args any extra arguments for the xml-rpc method
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(xml_method, *args)
            rc = @client.call(xml_method, *args)

            if !OpenNebula.is_error?(rc)
                @pe_id = rc
                rc     = nil
            end

            rc
        end

        # Calls to the corresponding update method to modify
        # the object's template
        #
        # @param [String] xml_method the name of the XML-RPC method
        # @param [String] new_template the new template contents
        # @param [Array] args any extra arguments for the xml-rpc method
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update(xml_method, new_template, *args)
            if new_template.nil?
                return Error.new('Wrong argument', Error::EXML_RPC_CALL)
            end

            call(xml_method, @pe_id, new_template, *args)
        end

        # Calls to the corresponding delete method to remove this element
        # from the OpenNebula core
        #
        # @param [String] xml_method the name of the XML-RPC method
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def delete(xml_method)
            call(xml_method, @pe_id)
        end

        # Calls to the corresponding chown method to modify
        # the object's owner and group
        #
        # @param [String] xml_method the name of the XML-RPC method
        # @param [Integer] uid the new owner id. Set to -1 to leave the current one
        # @param [Integer] gid the new goup id. Set to -1 to leave the current one
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chown(xml_method, uid, gid)
            call(xml_method, @pe_id, uid, gid)
        end

        # Calls to the corresponding chmod method to modify
        # the object's permission bits
        #
        # @param xml_method [String] the name of the XML-RPC method
        # @param octet [String] Permissions octed , e.g. 640
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod_octet(_xml_method, octet)
            owner_u = octet[0..0].to_i & 4 != 0 ? 1 : 0
            owner_m = octet[0..0].to_i & 2 != 0 ? 1 : 0
            owner_a = octet[0..0].to_i & 1 != 0 ? 1 : 0
            group_u = octet[1..1].to_i & 4 != 0 ? 1 : 0
            group_m = octet[1..1].to_i & 2 != 0 ? 1 : 0
            group_a = octet[1..1].to_i & 1 != 0 ? 1 : 0
            other_u = octet[2..2].to_i & 4 != 0 ? 1 : 0
            other_m = octet[2..2].to_i & 2 != 0 ? 1 : 0
            other_a = octet[2..2].to_i & 1 != 0 ? 1 : 0

            chmod(owner_u, owner_m, owner_a, group_u, group_m, group_a, other_u,
                  other_m, other_a)
        end

        # Calls to the corresponding chmod method to modify
        # the object's permission bits
        # Each [Integer] parameter must be 1 to allow, 0 deny, -1 do not change
        #
        # @param xml_method [String] the name of the XML-RPC method
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod(xml_method, owner_u, owner_m, owner_a, group_u, group_m, group_a, other_u,
                  other_m, other_a)
            call(xml_method, @pe_id, owner_u, owner_m,
                 owner_a, group_u, group_m, group_a, other_u,
                 other_m, other_a)
        end

        # Retrieves this Element's monitoring data from OpenNebula
        #
        # @param [String] xml_method the name of the XML-RPC method
        # @param xpaths [Array<String>] Xpath expressions for the elements to
        # retrieve.
        #
        # @return [Hash<String, Array<Array<int>>, OpenNebula::Error] Hash with
        #   the requested xpath expressions, and an Array of [timestamp, value].
        def monitoring(xml_method, xpaths)
            return Error.new('ID not defined') unless @pe_id

            rc = @client.call(xml_method, @pe_id)

            if OpenNebula.is_error?(rc)
                return rc
            end

            xmldoc = XMLElement.new
            xmldoc.initialize_xml(rc, 'MONITORING_DATA')

            OpenNebula.process_monitoring(xmldoc, @pe_id, xpaths)
        end

        public

        # Creates new element specifying its id
        # id:: identifyier of the element
        # client:: initialized OpenNebula::Client object
        def self.new_with_id(id, client = nil)
            new(build_xml(id), client)
        end

        # Returns element identifier
        # [return] _Integer_ the PoolElement ID
        def id
            @pe_id
        end

        # Gets element name
        # [return] _String_ the PoolElement name
        attr_reader :name

        # DO NOT USE - ONLY REXML BACKEND
        def to_str
            str = ''
            REXML::Formatters::Pretty.new(1).write(@xml, str)

            str
        end

        # Replace the xml pointed by xpath using  a Hash object
        # one object will be modified taking hash object pairs
        #
        # @param [String] xpath
        # @param [Hash] options object containing pair key-value
        #
        # @returns the new xml representation
        def replace(opts, xpath = 'TEMPLATE')
            return unless self[xpath]

            opts.each do |att, value|
                xpath_u = xpath+"/#{att}"
                docs = retrieve_xmlelements(xpath_u)
                if docs.size == 1
                    docs[0].set_content(value)
                end
            end
            update(template_like_str(xpath))
        end

    end

    # Processes the monitoring data in XML returned by OpenNebula
    #
    # @param [XMLElement] xmldoc monitoring data returned by OpenNebula
    #   monitorization timestamp
    # @param [Integer] oid Id of the object to process
    # @param [Array<String>] xpath_expressions Elements to retrieve.
    #
    # @return [Hash<String, Array<Array<int>>, OpenNebula::Error] Hash with
    #   the requested xpath expressions, and an Array of [timestamp, value].
    def self.process_monitoring(xmldoc, oid, xpath_expressions)
        hash = {}
        timestamps = xmldoc.retrieve_elements(
            "/MONITORING_DATA/MONITORING[ID=#{oid}]/TIMESTAMP"
        )

        xpath_expressions.each do |xpath|
            xpath_values = xmldoc.retrieve_elements(
                "/MONITORING_DATA/MONITORING[ID=#{oid}]/#{xpath}"
            )

            if xpath_values.nil?
                hash[xpath] = []
            else
                hash[xpath] = timestamps.zip(xpath_values)
            end
        end

        hash
    end

end
